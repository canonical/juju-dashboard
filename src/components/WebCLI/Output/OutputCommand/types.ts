import type { HistoryItem } from "store/juju/types";

import type { ProcessOutput, TableLinks } from "../../Output/types";

export type Props = {
  processOutput?: ProcessOutput;
  tableLinks?: TableLinks;
} & HistoryItem;
