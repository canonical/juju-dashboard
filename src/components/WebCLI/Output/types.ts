import type { ReactNode } from "react";

import type { HistoryItem } from "store/juju/types";

export type Block = {
  header: string;
  rows: string[];
};

export type Header = {
  content: string;
  ansiTitle: string;
  title: string;
  start: number;
  end?: number;
};

export type Column = {
  content: string;
  key: string;
  ansiValue: string;
  value: string;
  whitespaceBefore?: string;
  whitespaceAfter?: string;
};

export type TableLinksLink =
  | {
      externalLink: string;
    }
  | {
      link: string;
    };

type TableLinksGenerateLink = (column: Column, row: Column[]) => TableLinksLink;

type TableLinksBlocks = Record<string, TableLinksGenerateLink>;

export type CommandHandler = {
  /**
   * Wether the command needs to exactly match e.g. should it match `status` or
   * also match `status postgres`
   */
  exact?: boolean;
};

export type TableLinksCommand = {
  blocks: Record<string, TableLinksBlocks>;
} & CommandHandler;

export type TableLinks = Record<string, TableLinksCommand>;

export type ProcessCommand = {
  process: (messages: string[]) => ReactNode;
} & CommandHandler;

export type ProcessOutput = Record<string, ProcessCommand>;

export type Props = {
  content: HistoryItem[];
  helpMessage: ReactNode;
  loading?: boolean;
  processOutput?: null | ProcessOutput;
  tableLinks?: null | TableLinks;
  showHelp: boolean;
  setShouldShowHelp: (showHelp: boolean) => void;
};

export enum TestId {
  CODE = "output-code",
  CONTENT = "output-content",
  HELP = "output-help",
}
