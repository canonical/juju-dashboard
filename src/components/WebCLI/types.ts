export enum Label {
  CONNECTION_ERROR = "Unable to connect to the model.",
  AUTHENTICATION_ERROR = "Unable to authenticate.",
  NOT_OPEN_ERROR = "WebSocket connection is not open.",
  UNKNOWN_ERROR = "Unknown error.",
  HELP = "Display help",
}

export enum Fields {
  COMMAND = "command",
}
