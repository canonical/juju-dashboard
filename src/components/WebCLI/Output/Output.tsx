import { useListener, usePrevious } from "@canonical/react-components";
import fastDeepEqual from "fast-deep-equal/es6";
import type { FC } from "react";
import { useCallback, useEffect, useRef, useState } from "react";

import OutputCommand from "./OutputCommand";
import {
  HELP_HEIGHT,
  CONSIDER_CLOSED,
  DEFAULT_HEIGHT,
  AUTO_SCROLL_DISTANCE,
} from "./consts";
import type { Props } from "./types";
import { TestId } from "./types";

const dragHandles = ["webcli__output-dragarea", "webcli__output-handle"];

const WebCLIOutput: FC<Props> = ({
  content,
  helpMessage,
  loading = false,
  processOutput,
  showHelp,
  setShouldShowHelp,
  tableLinks,
}: Props) => {
  const resizeDeltaY = useRef(0);
  const dragNode = useRef<HTMLDivElement>(null);
  const outputRef = useRef<HTMLPreElement>(null);
  const [height, setHeight] = useState(1);
  const [dragHeight, setDragHeight] = useState(0);
  const showHelpPrevious = usePrevious(showHelp, false);
  const contentPrevious = usePrevious(content, false);
  const contentChanged = !fastDeepEqual(content, contentPrevious);

  const onResize = useCallback(() => {
    if (dragNode.current) {
      setDragHeight(dragNode.current.clientHeight);
    }
  }, []);

  useEffect(() => {
    if (contentChanged && outputRef.current) {
      const newResponseHeight =
        document.querySelector(".webcli__output-content-response:last-child")
          ?.clientHeight ?? 0;
      const { scrollHeight } = outputRef.current;
      // Get the distance the scroll area is from the bottom.
      const distance =
        scrollHeight -
        (outputRef.current.clientHeight +
          outputRef.current.scrollTop +
          // Need to get the position from the bottom before the new content in.
          newResponseHeight);
      if (distance <= AUTO_SCROLL_DISTANCE) {
        outputRef.current.scrollTo({ top: scrollHeight });
      }
    }
  }, [contentChanged]);

  useListener(window, onResize, "resize", true, true);

  useEffect(() => {
    // Set the drag height on first render.
    onResize();
  }, [onResize]);

  useEffect(() => {
    const resize = (clientY: number): void => {
      const viewPortHeight = window.innerHeight;
      const mousePosition = clientY + 40; // magic number
      const newHeight = viewPortHeight - mousePosition + resizeDeltaY.current;
      const maximumOutputHeight = viewPortHeight - 100; // magic number.
      if (newHeight < maximumOutputHeight && newHeight >= 10) {
        setHeight(newHeight);
      }
    };

    const resizeMouse = (ev: MouseEvent): void => {
      ev.preventDefault();
      resize(ev.clientY);
    };

    const resizeTouch = (ev: TouchEvent): void => {
      ev.preventDefault();
      resize(ev.touches[0].clientY);
    };

    const addMouseResizeListener = (ev: MouseEvent): void => {
      const target = ev.target as HTMLElement;
      if (!dragHandles.includes(target?.classList.value)) {
        return;
      }
      resizeDeltaY.current = ev.offsetY ?? 0;
      document.addEventListener("mousemove", resizeMouse);
    };

    const addTouchResizeListener = (ev: TouchEvent): void => {
      const target = ev.target as HTMLElement;
      if (!dragHandles.includes(target?.classList.value)) {
        return;
      }
      const rect = target.getBoundingClientRect();
      if (typeof ev.targetTouches?.[0]?.pageY === "number") {
        resizeDeltaY.current = ev.targetTouches[0].pageY;
      } else if (typeof rect?.top === "number") {
        resizeDeltaY.current = 0 - rect.top;
      } else {
        resizeDeltaY.current = 0;
      }
      document.addEventListener("touchmove", resizeTouch, { passive: false });
    };

    const removeListener = (): void => {
      document.removeEventListener("mousemove", resizeMouse);
      document.removeEventListener("touchmove", resizeTouch);
    };

    window.addEventListener("mousedown", addMouseResizeListener);
    window.addEventListener("mouseup", removeListener);
    window.addEventListener("touchstart", addTouchResizeListener, {
      passive: false,
    });
    window.addEventListener("touchend", removeListener, { passive: false });
    return (): void => {
      window.removeEventListener("mousedown", addMouseResizeListener);
      window.removeEventListener("mouseup", removeListener);
      window.removeEventListener("touchstart", addTouchResizeListener);
      window.removeEventListener("touchend", removeListener);
    };
  }, []);

  useEffect(() => {
    if (showHelp) {
      setHeight(HELP_HEIGHT + dragHeight);
    } else if (!showHelp && showHelpPrevious) {
      // The help got closed so also close the panel.
      setHeight(0);
    }
  }, [showHelp, dragHeight, showHelpPrevious]);

  useEffect(() => {
    if (height < CONSIDER_CLOSED) {
      setShouldShowHelp(false);
    }
  }, [height, setShouldShowHelp]);

  useEffect(() => {
    // New content is coming in, so check if we're collapsed and if we
    // are then open it back up.
    if (
      // Only trigger this condition if the content changes.
      contentChanged &&
      content.length &&
      height <= HELP_HEIGHT &&
      height !== DEFAULT_HEIGHT
    ) {
      setHeight(DEFAULT_HEIGHT);
      outputRef.current?.scrollTo({ top: outputRef.current.scrollHeight });
    }
  }, [content, contentChanged, height]);

  const output = content.map((historyItem, i) => (
    <OutputCommand
      {...historyItem}
      key={`message-${i}`}
      tableLinks={tableLinks}
      processOutput={processOutput}
    />
  ));

  return (
    <div className="webcli__output" style={{ height: `${height}px` }}>
      <div
        className="webcli__output-dragarea"
        aria-hidden={true}
        ref={dragNode}
      >
        <div className="webcli__output-handle"></div>
      </div>
      <pre
        className="webcli__output-content"
        style={{ height: `${height}px` }}
        data-testid={TestId.CONTENT}
        ref={outputRef}
      >
        {showHelp || (!loading && !content?.length) ? (
          <code data-testid={TestId.HELP}>{helpMessage}</code>
        ) : (
          <code data-testid={TestId.CODE}>{output}</code>
        )}
      </pre>
    </div>
  );
};

export default WebCLIOutput;
