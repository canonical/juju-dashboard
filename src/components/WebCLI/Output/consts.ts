export const DEFAULT_HEIGHT = 300;
// 20 is a magic number, sometimes the browser stops firing the drag at
// an inopportune time and the element isn't left completely closed.
export const CONSIDER_CLOSED = 20;
export const HELP_HEIGHT = 40;
