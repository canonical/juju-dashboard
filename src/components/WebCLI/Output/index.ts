export { default } from "./Output";
export { type Props as OutputProps, TestId as OutputTestId } from "./types";
