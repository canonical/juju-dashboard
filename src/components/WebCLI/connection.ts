import { toErrorString } from "utils";

export enum Label {
  INCORRECT_DATA_ERROR = "Invalid response from the server.",
}

type Options = {
  address: string;
  messageCallback: (messages: string[]) => void;
  onclose: () => void;
  onopen: () => void;
  onerror: (error: Event | string) => void;
  setLoading: (loading: boolean) => void;
};

class Connection {
  #address: string;
  #messageBuffer: string[] = [];
  #messageCallback: (messages: string[]) => void;
  #setLoading: Options["setLoading"];
  #wsOnOpen: () => void;
  #wsOnClose: () => void;
  #wsOnError: (error: Event | string) => void;
  #ws: null | WebSocket = null;

  constructor(options: Options) {
    this.#messageCallback = options.messageCallback;
    this.#address = options.address;
    this.#setLoading = options.setLoading;
    this.#wsOnOpen = options.onopen;
    this.#wsOnClose = options.onclose;
    this.#wsOnError = options.onerror;
  }

  get address() {
    return this.#address;
  }

  connect(onOpen?: () => void, onError?: (event: Event) => void) {
    const ws = new WebSocket(this.#address);
    ws.onopen = () => {
      this.#wsOnOpen();
      onOpen?.();
    };
    ws.onclose = this.#wsOnClose;
    ws.onmessage = this.#handleMessage.bind(this);
    ws.onerror = (event) => {
      this.#wsOnError(event);
      onError?.(event);
    };
    this.#ws = ws;
    return this;
  }

  send(message: string) {
    this.#setLoading(true);
    this.#ws?.send(message);
  }

  disconnect() {
    if (this.isActive()) {
      this.#ws?.close();
      this.#setLoading(false);
    }
    this.#messageBuffer = [];
  }

  reconnect() {
    return new Promise<void>((resolve, reject) => {
      this.connect(resolve, reject);
    });
  }

  isOpen() {
    return this.#ws?.readyState === WebSocket.OPEN;
  }

  isActive() {
    return this.#ws?.readyState === WebSocket.CONNECTING || this.isOpen();
  }

  isWebSocketEqual(newConnection: Connection) {
    return this.#ws === newConnection.#ws;
  }

  #handleMessage(messageEvent: MessageEvent) {
    this.#setLoading(false);
    try {
      let data: unknown;
      try {
        data = JSON.parse(messageEvent.data);
      } catch (error) {
        throw new Error(Label.INCORRECT_DATA_ERROR);
      }
      if (data === null || typeof data !== "object") {
        throw new Error(Label.INCORRECT_DATA_ERROR);
      }
      if ("done" in data && Boolean(data.done)) {
        // This is the last message so send the entire message.
        this.#messageCallback(this.#messageBuffer);
        this.#messageBuffer = [];
        return;
      }
      const dataOutput =
        "output" in data && Boolean(data.output) ? data.output : null;
      if (dataOutput === null || !dataOutput) {
        // This is the first message, an empty object and a newline.
        return;
      }
      if (!Array.isArray(dataOutput)) {
        throw new Error(Label.INCORRECT_DATA_ERROR);
      }
      // Build up messages until it gets the 'done' message:
      this.#messageBuffer.push(dataOutput[0]);
    } catch (error) {
      this.#wsOnError(toErrorString(error));
    }
  }
}

export default Connection;
