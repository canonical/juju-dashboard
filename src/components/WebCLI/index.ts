export { default } from "./WebCLI";
export {
  Label as WebCLILabel,
  Fields as WebCLIFields,
  TestId as WebCLITestId,
} from "./types";
