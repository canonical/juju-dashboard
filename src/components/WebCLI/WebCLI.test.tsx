import { fireEvent, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { WS } from "jest-websocket-mock";
import type { MockInstance } from "vitest";
import { vi } from "vitest";

import bakery from "juju/bakery";
import { generalStateFactory, configFactory } from "testing/factories/general";
import {
  commandHistoryState,
  commandHistoryItem,
} from "testing/factories/juju/juju";
import { rootStateFactory } from "testing/factories/root";
import { renderComponent } from "testing/utils";

import { OutputTestId } from "./Output";
import WebCLI from "./WebCLI";
import { Label as ConnectionLabel } from "./connection";
import { MAX_HISTORY } from "./consts";
import { Label } from "./types";

vi.mock("juju/bakery", () => ({
  __esModule: true,
  default: {
    storage: {
      get: vi.fn(),
    },
  },
}));

describe("WebCLI", () => {
  let bakerySpy: MockInstance;
  let server: WS;
  const props = {
    controllerWSHost: "localhost:1234",
    modelUUID: "abc123",
    credentials: {
      user: "eggman@external",
      password: "somelongpassword",
    },
    onCommandSent: vi.fn(),
    activeUser: "eggman@external",
  };

  beforeEach(() => {
    bakerySpy = vi.spyOn(bakery.storage, "get");
    server = new WS("wss://localhost:1234/model/abc123/commands");
  });

  afterEach(() => {
    bakerySpy.mockClear();
    localStorage.clear();
    WS.clean();
  });

  it("renders correctly", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    expect(document.querySelector(".webcli")).toMatchSnapshot();
  });

  it("displays preloaded history", async () => {
    const messages = [
      "Model         Controller",
      "test-model    localhost-localhost",
    ];
    const commandHistory = commandHistoryState.build({
      abc123: [
        commandHistoryItem.build({
          command: "status",
          messages: messages,
        }),
      ],
    });
    renderComponent(<WebCLI {...props} history={commandHistory} />);
    expect(await screen.findByTestId(OutputTestId.CONTENT)).toHaveTextContent(
      ["$ status", messages.join("\n")].join(""),
      { normalizeWhitespace: false },
    );
  });

  it("shows the help in the output when the ? is clicked", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    await userEvent.click(screen.getByRole("button"));
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(
      `Welcome to the Juju Web CLI - see the full documentation here.`,
    );
  });

  it("shows the help in the output when enter is pressed", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status{enter}");
    const messages = [{ output: ["test"] }, { done: true }];
    messages.forEach((message) => {
      server.send(JSON.stringify(message));
    });
    await waitFor(() =>
      expect(screen.getByTestId(OutputTestId.CONTENT)).toHaveTextContent(
        "test",
      ),
    );
    fireEvent.keyDown(screen.getByRole("button", { name: Label.HELP }), {
      key: "Enter",
    });
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(
      `Welcome to the Juju Web CLI - see the full documentation here.`,
    );
  });

  it("shows the help in the output when space is pressed", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    fireEvent.keyDown(screen.getByRole("button", { name: Label.HELP }), {
      key: " ",
    });
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(
      `Welcome to the Juju Web CLI - see the full documentation here.`,
    );
  });

  it("does not show the help in the output when keys other than space or enter are pressed", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    fireEvent.keyDown(screen.getByRole("button", { name: Label.HELP }), {
      key: "a",
    });
    expect(
      document.querySelector(".webcli__output-content code"),
    ).not.toHaveTextContent("");
  });

  it("shows the help when there is no output", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(
      `Welcome to the Juju Web CLI - see the full documentation here.`,
    );
  });

  it("trims the command being submitted", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "      status       {enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status"],
      }),
    );
  });

  it("navigate back through the history", async () => {
    localStorage.setItem(
      "cliHistory",
      JSON.stringify(["status", "help", "whoami"]),
    );
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "{arrowup}{arrowup}");
    expect(input).toHaveValue("help");
  });

  it("navigate forward through the history", async () => {
    localStorage.setItem(
      "cliHistory",
      JSON.stringify(["status", "help", "whoami"]),
    );
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "{arrowup}{arrowup}{arrowup}");
    expect(input).toHaveValue("status");
    await userEvent.type(input, "{arrowdown}{arrowdown}");
    expect(input).toHaveValue("whoami");
  });

  it("can navigate forward to the empty state", async () => {
    localStorage.setItem(
      "cliHistory",
      JSON.stringify(["status", "help", "whoami"]),
    );
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "{arrowup}{arrowup}");
    expect(input).toHaveValue("help");
    await userEvent.type(input, "{arrowdown}{arrowdown}");
    expect(input).toHaveValue("");
  });

  it("prevents navigating past the last history item", async () => {
    localStorage.setItem(
      "cliHistory",
      JSON.stringify(["status", "help", "whoami"]),
    );
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "{arrowup}{arrowup}{arrowup}{arrowup}");
    expect(input).toHaveValue("status");
  });

  it("limits the number of items in the history", async () => {
    const items = [...Array(MAX_HISTORY).keys()].map((_, i) => `command-${i}`);
    localStorage.setItem("cliHistory", JSON.stringify(items));
    // ..until it receives a 'done' message.
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    let history = JSON.parse(localStorage.getItem("cliHistory") ?? "");
    expect(history).toHaveLength(MAX_HISTORY);
    expect(history[0]).toBe("command-0");
    await userEvent.type(input, "status{enter}");
    history = JSON.parse(localStorage.getItem("cliHistory") ?? "");
    expect(history).toHaveLength(MAX_HISTORY);
    // The first item should be now be the old second item.
    expect(history[0]).toBe("command-1");
  });

  it("supports macaroon based authentication", async () => {
    const state = rootStateFactory.build({
      general: generalStateFactory.build({
        config: configFactory.build({
          controllerAPIEndpoint: "ws://localhost:1234/api",
        }),
        controllerConnections: {
          "wss://localhost:1234/api": {
            user: {
              "display-name": "eggman",
              identity: "user-eggman@external",
              "controller-access": "",
              "model-access": "",
            },
          },
        },
      }),
    });
    bakerySpy.mockImplementation((key) => {
      const macaroons: Record<string, string> = {
        "wss://localhost:1234/api": "WyJtYWMiLCAiYXJvb24iXQo=",
      };
      return macaroons[key];
    });

    renderComponent(<WebCLI {...props} credentials={null} />, { state });
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "      status       {enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        macaroons: [["mac", "aroon"]],
        commands: ["status"],
      }),
    );
  });

  it("sends commands over the websocket", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "some-command{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["some-command"],
      }),
    );
    expect(props.onCommandSent).toBeCalledWith("some-command");
  });

  it("displays messages received over the websocket", async () => {
    // ..until it receives a 'done' message.
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );

    const messages = [
      {
        output: [
          "Model       Controller       Cloud/Region     Version    SLA          Timestamp",
        ],
      },
      {
        output: [
          "controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z",
        ],
      },
      { output: [""] },
      {
        output: [
          "Machine  State    DNS             Inst id        Series  AZ          Message",
        ],
      },
      {
        output: [
          "0        started  35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING",
        ],
      },
      { output: [""] },
      { done: true },
    ];

    messages.forEach((message) => {
      server.send(JSON.stringify(message));
    });

    const code = await screen.findByTestId(OutputTestId.CODE);
    expect(code?.textContent).toMatchSnapshot();
    // Wait for the setTimeout that buffers the message updates in connection.js.
    await waitFor(() => {
      expect(screen.getByTestId(OutputTestId.CONTENT)).toHaveAttribute(
        "style",
        "height: 300px;",
      );
    });
  });

  it("calls the onHistoryChange callback when it receives messages", async () => {
    const onHistoryChange = vi.fn();
    renderComponent(<WebCLI {...props} onHistoryChange={onHistoryChange} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status{enter}");
    const messages = [
      "Model         Controller           Cloud/Region         Version    SLA          Timestamp",
      "test-model    localhost-localhost  localhost/localhost  3.2-beta3  unsupported  01:17:46Z",
    ];
    messages.forEach((message) => {
      server.send(JSON.stringify({ output: [message] }));
    });
    server.send(JSON.stringify({ done: true }));
    expect(onHistoryChange).toHaveBeenCalledWith("abc123", {
      command: "status",
      messages,
    });
  });

  it("displays ansi colored content colored", async () => {
    // ..until it receives a 'done' message.
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );

    const messages = [
      {
        output: [
          "Model       Controller       Cloud/Region     Version    SLA          Timestamp",
        ],
      },
      {
        output: [
          "controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z",
        ],
      },
      { output: [""] },
      {
        output: [
          "Machine  State    DNS             Inst id        Series  AZ          Message",
        ],
      },
      {
        output: [
          "0        \u001b[32mstarted  \u001b[0m35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING",
        ],
      },
      { output: [""] },
      { done: true },
    ];

    messages.forEach((message) => {
      server.send(JSON.stringify(message));
    });

    const code = await screen.findByTestId(OutputTestId.CODE);
    expect(code?.textContent).toMatchSnapshot();
    // Wait for the setTimeout that buffers the message updates in connection.js.
    await waitFor(() => {
      expect(screen.getByTestId(OutputTestId.CONTENT)).toHaveAttribute(
        "style",
        "height: 300px;",
      );
    });
  });

  it("can be resized with a mouse", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    expect(await screen.findByTestId(OutputTestId.CONTENT)).toHaveAttribute(
      "style",
      "height: 0px;",
    );
    const handle = document.querySelector(".webcli__output-handle");
    expect(handle).toBeTruthy();
    if (handle) {
      fireEvent.mouseDown(handle);
      fireEvent.mouseMove(handle, {
        clientY: 100,
      });
    }
    expect(await screen.findByTestId(OutputTestId.CONTENT)).toHaveAttribute(
      "style",
      "height: 628px;",
    );
  });

  it("can be resized on mobile", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    expect(await screen.findByTestId(OutputTestId.CONTENT)).toHaveAttribute(
      "style",
      "height: 0px;",
    );
    const handle = document.querySelector(".webcli__output-handle");
    expect(handle).toBeTruthy();
    if (handle) {
      fireEvent.touchStart(handle);
      fireEvent.touchMove(handle, {
        touches: [
          {
            clientY: 100,
          },
        ],
      });
    }
    expect(await screen.findByTestId(OutputTestId.CONTENT)).toHaveAttribute(
      "style",
      "height: 628px;",
    );
  });

  it("should display connection error when no websocket address is present", () => {
    renderComponent(<WebCLI {...props} modelUUID="" />);
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${Label.CONNECTION_ERROR}`);
  });

  it("should display authentication error when no credentials and macaroons are available", async () => {
    bakerySpy.mockImplementation(() => null);

    renderComponent(<WebCLI {...props} credentials={null} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({ commands: ["status"] }),
    );
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${Label.AUTHENTICATION_ERROR}`);
  });

  it("should display error when JSON can't be parsed", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );
    server.send("[1, 2, 3, 4}");
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${ConnectionLabel.INCORRECT_DATA_ERROR}`);
  });

  it("should display error when received message is falsy", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );
    server.send(JSON.stringify(null));
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${ConnectionLabel.INCORRECT_DATA_ERROR}`);
  });

  it("should display error when received message is not an object", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );
    server.send(JSON.stringify("not an object"));
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${ConnectionLabel.INCORRECT_DATA_ERROR}`);
  });

  it("should display error when connecting to sub controller of JAAS", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );
    new WS("invalid-url-for-jaas-sub-controller");
    server.send(
      JSON.stringify({ "redirect-to": "invalid-url-for-jaas-sub-controller" }),
    );
    await server.closed;
    // Trying to connect to WebSocket with invalid URL spits up error.
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${ConnectionLabel.JAAS_CONNECTION_ERROR}`);
  });

  it("should display error when received message output is not an array", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    await expect(server).toReceiveMessage(
      JSON.stringify({
        user: "eggman@external",
        credentials: "somelongpassword",
        commands: ["status --color"],
      }),
    );
    server.send(
      JSON.stringify({
        output: "not-an-array",
      }),
    );
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${ConnectionLabel.INCORRECT_DATA_ERROR}`);
  });

  it("should display unknown errors", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    server.error();
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${Label.UNKNOWN_ERROR}`);
  });

  it("should display error when trying to send message over an unopened connection", async () => {
    renderComponent(<WebCLI {...props} />);
    await server.connected;
    server.close();
    const input = screen.getByRole("textbox");
    await userEvent.type(input, "status --color{enter}");
    expect(
      document.querySelector(".webcli__output-content code"),
    ).toHaveTextContent(`ERROR: ${Label.NOT_OPEN_ERROR}`);
  });
});
