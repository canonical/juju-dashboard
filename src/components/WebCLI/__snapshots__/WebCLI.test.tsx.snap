// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebCLI > displays ansi colored content colored 1`] = `
"$ status --colorModel       Controller       Cloud/Region     Version    SLA          Timestamp
controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z

Machine  State    DNS             Inst id        Series  AZ          Message
0        started  35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING

"
`;

exports[`WebCLI > displays messages received over the websocket 1`] = `
"$ status --colorModel       Controller       Cloud/Region     Version    SLA          Timestamp
controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z

Machine  State    DNS             Inst id        Series  AZ          Message
0        started  35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING

"
`;

exports[`WebCLI > renders correctly 1`] = `
<div
  class="webcli is-dark"
  data-testid="WebCLI"
>
  <div
    class="webcli__output"
    style="height: 0px;"
  >
    <div
      aria-hidden="true"
      class="webcli__output-dragarea"
    >
      <div
        class="webcli__output-handle"
      />
    </div>
    <pre
      class="webcli__output-content"
      data-testid="output-content"
      style="height: 0px;"
    >
      <code
        data-testid="output-help"
      >
        Welcome to the Juju Web CLI - see the
         
        <a
          href="https://documentation.ubuntu.com/juju/latest/reference/juju-web-cli/"
          rel="noreferrer"
          target="_blank"
        >
          full documentation here
        </a>
        .
      </code>
    </pre>
  </div>
  <div
    class="webcli__input"
  >
    <div
      class="webcli__input-prompt"
    >
      $ juju
    </div>
    <form>
      <input
        aria-label="enter command"
        autocapitalize="none"
        autocomplete="off"
        autocorrect="off"
        class="webcli__input-input"
        name="command"
        placeholder="enter command"
        spellcheck="false"
        type="text"
      />
    </form>
    <div
      class="webcli__input-help"
    >
      <i
        aria-label="Display help"
        class="p-icon--help is-light"
        role="button"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
